java/arcs/core/util/Scheduler.kt

```diff
@@ -122,7 +121,7 @@ class Scheduler(
     /** Wait for the [Scheduler] to become idle. */
     /* internal */
     suspend fun waitForIdle() {
-        idlenessFlow.debounce(50).filter { it }.first()
+        idlenessFlow.filter { it }.first()
```

Comment:
The reason I was debouncing here was to try and smooth out situations where tasks are frequently scheduled with the Scheduler. Without debouncing, it's almost useless.. and instead signifies only that we are between scheduler agendas.

Reply:
I understand we might be in between scheduler agendas.
What I observed is it's already idle while calling waitForIdle and the debounce always sets a 50ms timer and waits for the timer expiration then returns the idleness, thus the entire system does nothing but waits.
We cannot foresee when a new agenda will be queued and at what frequency.
If the debouncing time is really required, can we reduce it from 50 ms to a subframe time e.g. 5 ms for 90fps after all the scheduler serves client apis and callbacks and would be better to response quickly instead of sleeping there?

Another idea for future optimization could be: adaptive debouncing timer i.e.
a new agenda adds a X ms to the debouncing timer, when all agendas are finished, the debouncing timer starts to decay gradually and eventually down to 0.

Reply:
What do you think of this idea?  When the idlenessFlow goes to `true`, we stash a timestamp... then in `waitForIdle` we first check the current state of the idleness value. 

* If it's `true` we check the timestamp against the current time and return if the timestamp is "old enough".
* If it's `false` we start the debounce flow consumer.

Reply:
How should we define a balanced "old enough" time?
Does using 5 or 10 ms make sense to you?

Considering a simple tweak to your proposal as a starting point like:
"old enough": 10 ms
- if it's `true` and bigger than 10ms, return
- if it's `true` but less than 10ms, start a (10ms - (now() - stashed timestamp)) debounce flow consumer.
  - if it's `true` after the flow's `first()`, return  // this gives a quicker turn-around
  - else, start a 10ms debounce flow consumer
- else, start a 10ms debounce flow consumer.

Reply:
I wonder if the whole 'waitForIdle' paradigm is wrong, and we should really have a 'runDuringIdle(block: () -> Unit)' function. Are we waiting for commits to occur with waitForIdle(), or trying to avoid races?

Reply:
The `waitForIdle()` is for `client` to wait until all scheduler agendas are handled.
There are no pending agendas at any point of idleness.

The debounce was to try and smooth out situations where tasks are frequently scheduled with the Scheduler as Jason mentioned i.e. a tons of updates coming and interleaving in a short interval, but the debouncing time needs to be well (re)considered at least not impacting user experience after all schedulers take care of client apis and callbacks and should not become bottleneck during user interactions.

Reply:
@jasonwyatt ptal at the new change which applies the logic aforementioned.